feat: also include errored specfiles when using sharding

FinKingma opened the pull request:
hey,

Now as long as the output contains sharding + Spec: description, protractor-flake will also re-run the test files that went into error state.

Problem:
In the situation that:
- Test A has a failed expectation
- Test B went into error state
- Test C passed
  Only Test A will be re-run. And if the test passes, the whole run will have passed (marked unstable)
  But Test B has never been run, it has been skipped.

Fix:
Now in the situation that Specs: is described, and you use sharing, all the test files that did NOT return an OK signal, will be rerun - instead of the tests that returned with a failure.

Commit:
feat: also include errored specfiles when using sharding

NickTomlin commented:
Thanks for putting this together! I'm currently on limited internet access for the next week and a half but I will give it a look after that.

ayush commented:
@NickTomlin did you get a chance to review this PR? Looks like it would fix https://github.com/NickTomlin/protractor-flake/issues/21 as well.

NickTomlin commented:
Sorry for the delay. I think this feature is 👍  but I'd like to switch up a few things before we merge it. `failed-spec-parsers` is starting to get a little messy, so i'd like to do a little reorganization. 

I spiked out something in [this branch](https://github.com/NickTomlin/protractor-flake/commit/d96d3cdb89a401a33291d95e16237b28517bc0c0) which is the direction i'd like to head. 

What do y'all think?

NickTomlin commented on src/failed-spec-parser.js:
```diff
@@ -2,8 +2,34 @@ export default function (output = '') {
   let match = null
   let CUCUMBERJS_TEST = /^\d+ scenarios?/m
   let failedSpecs = new Set()
+  let PROTRACTOR_SHARDED = /------------------------------------/g
```
I believe using multiple browsers also outputs these lines. I'm not sure if there's another way of detecting this. We should double check that it doesn't interfere with non-sharded multi browser tests

NickTomlin commented on src/failed-spec-parser.js:
```diff
@@ -2,8 +2,34 @@ export default function (output = '') {
 
-  if (CUCUMBERJS_TEST.test(output)) {
+  if (PROTRACTOR_SHARDED.test(output) && SPECFILE_REG.test(output)) {
+    console.log('sharded test found')
```
I'd nix this log

NickTomlin commented on src/failed-spec-parser.js:
```diff
@@ -2,8 +2,34 @@ export default function (output = '') {
+  if (PROTRACTOR_SHARDED.test(output) && SPECFILE_REG.test(output)) {
+    console.log('sharded test found')
+    let testsOutput = output.split('------------------------------------')
+    testsOutput.shift()
```
Could we `slice` on the previous line instead of shifting here?

Commit:
added new test for multi capabilities

Commit:
reverted * to +

Commit:
rename sharded -> multi test

Commit:
simplified checking failures

FinKingma commented:
Looks good. I added a PR with the proposed changes and also added a new test for multi capabilities with different kind of words (that were outputted on my project, like specs instead of assertions / tests), I implemented another fix to make it work.

NickTomlin commented:
awesome, i'll take a look at this after work today. Thanks!

NickTomlin commented:
Awesome, I had a chance to test this a little and it's looking good. I'm getting a little confused by the multiple prs and commits so I propose we this:
1. Merge your changes into https://github.com/NickTomlin/protractor-flake/tree/reorganize-parsers
2. Iterate as needed
3. Merge the reorganize-parsers branch
4. Close this pr and profit 😄

FinKingma commented:
New PR created with all fixes and closed the other 2.
